src/ipips/ipip-0499.md

```diff
@@ -0,0 +1,131 @@
+
+Profiles explicitly define the UnixFS parameters, e.g. dag width, hash algorithem, and chunk size, that affect the resulting CID, such that given the profile and input data different implementations will generate identical CIDs.
+
+## Motivation
+
+UnixFS CIDs are not deterministic. This means that the same file tree can yield different CIDs depending on the parameters used by the implementation to generate it, which in some cases, aren't even configurable by the user. For example, the chunk size, DAG width, and layout can vary between implementations or even between different versions of the same implementation.
+
+This lack of determinism makes has a number of drawbacks:
+
+- It is difficult to verify content across different tools and implementations, as the same content may yield different CIDs.
+- Users are requires to store and transfer UnixFS merkle proofs in order to verify CIDs, adding storage overhead, network bandwidth, and complexity to the verification process.
+- In terms of developer experience, it deviates from the mental model of a hash function, where the same input should always yield the same output. This leads to confusion and frustration when working with UnixFS CIDs
+
+By introducing profiles, we can benefit from both the optionality offered by UnixFS, where users are free to chose their own parameters, and determinism through profiles.
+
+UnixFS CIDs can be generated with a variety of settings and optimizations for chunking, DAG width, and more. This means the same file tree can yield multiple, different CIDs depending on which tools and settings are used, and it is not possible to reliably reproduce or verify the CID. Profiles offer With profiles, following the same profile will produce identical CIDs for identical content, whic makes verification regardless of implementation.
+
+## Detailed design
+
+We introduce a set of named profiles that define a set of parameters for generating UnixFS CIDs. These profiles can be used by implementations to ensure that the same content will yield the same CID across different tools and implementations.
+
+### UnixFS parameters
+
+The profiles define a set of parameters that affect the resulting CID. These parameters are based on the UnixFS specification and are used to generate the CID for a given file tree. The parameters include:
+
+1. CID version (currently only CIDv0 or CIDv1)
+1. Hash function
+1. UnixFS chunk size
+1. UnixFS DAG layout (e.g. balanced, trickle)
+1. UnixFS DAG width (max number of links per `File` node)
+1. `HAMTDirectory` fanout (must be a power of 2)
+1. `HAMTDirectory` threshold (max `Directory` size before switching to `HAMTDirectory`): based on an estimate of the block size by counting the size of PNNode.Links
+1. Leaf Envelope: either `dag-pb` or `raw`
+1. Whether empty directories are included in the DAG
+  - Some implementations apply filtering before merkleizing filesystem entries in the DAG.
+
```

Comment:
couple of other things to consider?

* Directory wrapping at the top level (for just files, kubo has an option to wrap in a directory so you get file metadata)
* Presence and accurate setting of `Tsize` - at one point we were going to deprecate this field for some cases, although I think all our encoders now do it properly, you could just mandate this in the spec though -- all valid profiles must properly encode `Tsize`.

Reply:
I've added this as a parameter. 

According to the latest version of https://github.com/ipfs/specs/pull/331/files, the calculation is done as follows:

> To compute the Tsize of a child DAG, sum the length of the dag-pb outside message binary length and the blocksizes of all nodes in the child DAG.

If calculated according to this, does it make accurate?

Reply:
sounds about right, I remember there being some nuance in exactly what's included in the size calculation, making it not super stable if you get it slightly wrong (as we did for some variants in go-unixfsnode for a while)